Adts for sql tables

sashasashasasha151 opened the pull request:
## Description

This PR defines data types necessary to interact with PostgreSQL DB via `beam` and adds simple tests to check validity of these types.

## Related issue(s)

<!--
- Short description of how the PR relates to the issue, including an issue link.
For example
- Fixed #100500 by adding lenses to exported items

Write 'None' if there are no related issues (which is discouraged).
-->

https://issues.serokell.io/issue/EDNA-41

## :white_check_mark: Checklist for your Pull Request

<!--
Ideally a PR has all of the checkmarks set.

If something in this list is irrelevant to your PR, you should still set this
checkmark indicating that you are sure it is dealt with (be that by irrelevance).

If you don't set a checkmark (e. g. don't add a test for new functionality),
you must be able to justify that.
-->

#### Related changes (conditional)

- Tests
  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)

#### Stylistic guide (mandatory)

- [x] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Commit-and-PR-policy-4cf98e1a910a415d86b5f2491d9af1af).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

sashasashasasha151 committed:
Move db initialisation functions to separate module

sashasashasasha151 committed:
Create types for sql tables

Problem: to use beam we need to add corresponding types which will
present DB tables

Solution: create them

Limitation: there are no useful wrappers such as
`newtype Name = Name Text`.

gromakovsky commented on backend/src/Edna/DB/Schema.hs:
```diff
@@ -1,5 +1,325 @@
+data AnalysisMethodT f = AnalysisMethodRec
+  { amAnalysisMethodId :: C f (SqlSerial Int)
+  , amDescription :: C (Nullable f) Text
+  , amParameters :: C (Nullable f) (PgJSON String) -- TODO add corresponding type
```
This should probably be just `()` for now, we don't have any parameters yet.

gromakovsky commented on backend/src/Edna/DB/Schema.hs:
```diff
@@ -1,5 +1,325 @@
+  , seAnalysisMethodId :: C f Int
+  , seExperimentId :: C f Int
+  , seIsSuspicious :: C f Bool
+  , seResult :: C f (PgJSON String) -- TODO add corresponding type
```
I suppose it should be `Double` for now, just one IC50 value.

gromakovsky commented on backend/src/Edna/DB/Schema.hs:
```diff
@@ -1,5 +1,325 @@
+  { efExperimentFileId :: C f (SqlSerial Int)
+  , efMethodologyId :: C (Nullable f) Int
+  , efUploadDate :: C f UTCTime
+  , efMeta :: C (Nullable f) (PgJSON String) -- TODO add corresponding type
```
I'll define a type to replace this `String`.

gromakovsky replied:
If we merge #21 first, it can be `FileMetadata` from that PR.

gromakovsky replied:
I've done it myself.

gromakovsky reviewed (commented):
Looks good so far, waiting for tests to check that all is fine and see how it all can be used.

sashasashasasha151 committed:
Add tests for DB schema

Problem: we'd like to check that previously added types for DB
interaction match `init.sql`.
Solution: add Hedgehog-based property tests where we generate
random items, insert them into DB, then read and check that they
are equal to the originally put values.

gromakovsky commented on backend/src/Edna/DB/Schema.hs:
```diff
@@ -1,5 +1,327 @@
+--------------------------
+
+data ProjectT f = ProjectRec
+  { pProjectId :: C f (SqlSerial Int32)
```
They say that its range is `1 to 2147483647`, so I think `Word32` is slightly more accurate here, even though `Int32` should work too.

sashasashasasha151 replied:
For all ids or how?

gromakovsky replied:
Yes, I meant all IDs. I don't insist on changing `Int32` to `Word32` everywhere, it's really minor, I just think it's slightly more correct/accurate. So it's up to you.

gromakovsky commented on backend/test/Test/DB/DBSpec.hs:
```diff
@@ -0,0 +1,107 @@
+spec = withContext $ describe "Validate schema" $ modifyMaxSuccess (const 20) $ do
+  let EdnaSchema {..} = ednaSchema
+  let insertWithConflict schema value = runInsert' $ Pg.insert schema value $
+        Pg.onConflict (Pg.conflictingFields primaryKey) Pg.onConflictUpdateAll
```
Do we expect conflicts in tests?

gromakovsky commented on backend/test/Test/DB/DBSpec.hs:
```diff
@@ -0,0 +1,107 @@
+import Test.Setup (ednaTestMode, withContext)
+
+spec :: Spec
+spec = withContext $ describe "Validate schema" $ modifyMaxSuccess (const 20) $ do
```
I think `min 20` is slightly better than `const 20`, because in this case one can at least decrease the number of test runs. So it is 20 by default, but if I want 5 for some reason, I can specify 5 and it will work.

gromakovsky commented on backend/test/Test/DB/DBSpec.hs:
```diff
@@ -0,0 +1,107 @@
+        Pg.onConflict (Pg.conflictingFields primaryKey) Pg.onConflictUpdateAll
+  let commonCheck schema expected = do
+        actual <- lift $ runSelectReturningOne' $ select $ all_ schema
+        liftIO $ actual `shouldBe` Just expected
```
The logic of tests is good, but there are some problems and I don't quite understand what's going on. I tried to debug it, but couldn't succeed yet. So let me explain.
I see that CI fails with the following error:

>            HUnitFailure (Just (SrcLoc {srcLocPackage = "main", srcLocModule = "Test.DB.DBSpec", srcLocFile = "test/Test/DB/DBSpec.hs", srcLocStartLine = 33, srcLocStartCol = 18, srcLocEndLine = 33, srcLocEndCol = 49})) (ExpectedButGot Nothing "Just (MeasurementRec {mMeasurementId = SqlSerial {unSerial = 1}, mExperimentId = 1, mConcentration = 73.06692048416038, mSignal = 24.241577931890433, mIsOutlier = False})" "Just (MeasurementRec {mMeasurementId = SqlSerial {unSerial = 1}, mExperimentId = 1, mConcentration = 73.0669204841604, mSignal = 24.2415779318904, mIsOutlier = False})")

For some reason `Double` values are read a bit imprecisely. I don't know why it happens, it works as expected on my machine. I didn't try to investigate this particular problem, I'm leaving it to you.

The problem that I see here is that the error message looks bad, it is raw `HUnitFailure`. Inside `MonadTest` you should use `===` from `Hedgehog`, it prints the error more nicely and even shows the diff. So you can write `actual === Just expected` here. But when I did it, I was surprised because I started getting a completely weird error:
```
           SqlError {sqlState = "23503", sqlExecStatus = FatalError, sqlErrorMsg = "insert or update on table \"experiment\" violates foreign key constraint \"has_experiment_file\"", sqlErrorDetail = "Key (experiment_file_id)=(1) is not present in table \"experiment_file\".", sqlErrorHint = ""}
```

FTR I did this to break one test:
```
  it "Measurement" $ \ctx -> ednaTestMode ctx $
    insertMeasurement >>= (\m -> commonCheck esMeasurement (m {mSignal = 200}))
```
and commented out all other tests.

I still don't understand why it happens. I tried debugging it, even changed `client_min_messages` that we pass to Postgres to see its logs, but I still don't understand.

One thing I noticed is that if I set `modifyMaxShrinks (const 1)` (or 2 or 3), I will get an error about mismatch between `mSignal` as expected. If max number of shrinks is 11, I consistently get that error. Shrinks are not very useful in these particular tests, so as a workaround we can do `modifyMaxShrinks (const 3)` for example. But I am really curious why it happens and maybe indicates that we are doing something wrong. I'll probably stop at this point, but maybe will make one more attempt later today. I'll be glad if you figure out why it happens, also maybe it happens only on machine :shrug: That said, I don't think we should spend too much time on it because we have a lot of more urgent work. But since you worked with Postgresql from Haskell more than me, maybe you'll figure it our quickly.

2 more points:
1. In `ednaTestMode` currently you init and reset schema without `bracket`. Perhaps it makes sense to use `bracket` there, because AFAIU we want to reset shema every time. Alternatively, we can run `resetSchema` _before_ `initSchema`, AFAIU it simply does nothing if schema is not initialized. Maybe it will also solve the problem above.
2. `Hedgehog` prints code around the place that fails. It's based on `HasCallStack` machinery. When I used `===` in `commonCheck` and caused it to fail, I got a very large error message with almost the whole body of `spec`. I think we don't that much code in error messages, and AFAIU it can be solved by explicitly defining the type of `commonCheck` and adding `HasCallStack` constraint to it.

gromakovsky replied:
I even tried `TEST_ARGUMENTS='-j1' make test`, but it also didn't help.

gromakovsky commented on backend/dev-config.yaml:
```diff
@@ -6,7 +6,7 @@ api:
   # Syntax: https://www.postgresql.org/docs/9.5/libpq-connect.html#LIBPQ-CONNSTRING
   # It should match parameters from `/deployment/docker-compose.yml`
-  conn-string: "host=localhost dbname=ednadb user=ednauser password=ednapass"
+  conn-string: "host=localhost dbname=edna user=postgres password=12345"
```
But why?

sashasashasasha151 replied:
Will remove. Pushed it accidentally

gromakovsky commented:
I have checked your fixes and squashed them.

gromakovsky commented on backend/test/Test/Setup.hs:
```diff
@@ -49,4 +49,7 @@ withContext = around withContext'
+        callback ctx
+
+ednaTestMode :: EdnaContext -> PropertyT Edna a -> PropertyT IO ()
+ednaTestMode ctx = void . hoist (\action -> runRIO ctx $ schemaInit *> action <* runPg resetSchema)
```
Also what do you think about my suggestion to wrap it into `bracket` or do `resetSchema` before `schemaInit`?

Also for SMT we also need to do `schemaInit` and `resetSchema`, but there we can't use `ednaTestMode`, so it would be nice to extract `schemaInit *> action <* runPg resetSchema` into another function.

sashasashasasha151 replied:
What is the difference between `schemaInit` before `resetSchema` and `resetSchema` before `schemaInit`? And how `bracket` solve this issue?

gromakovsky replied:
Currently if `action` throws an uncaught exception, `resetSchema` won't run. But I suppose this exception may be caught by testing engine afterwards (I am not sure if all exceptions are caught, but at least things like `HUnitFailure` should be caught for sure) and then the next test will run, but schema won't be reset when we run it.

If we use `bracket`, `resetSchema` will always happen. Or we can move it before `schemaInit` and then `resetSchema` will always happen if `schemaInit` happens.

sashasashasasha151 replied:
I made it in a little bit different way

gromakovsky commented on backend/src/Edna/DB/Schema.hs:
```diff
@@ -1,39 +1,331 @@
+data ExperimentT f = ExperimentRec
+  { eExperimentId :: C f (SqlSerial Int32)
+  , eExperimentFileId :: C f Int32
+  , eCompoundId :: C f Int32
```
![image](https://user-images.githubusercontent.com/2556409/111868855-68686c00-898d-11eb-8553-eb2c0f5f59a8.png)

Here they suggest using `PrimaryKey` to refer to primary keys of other tables. Shouldn't `eCompoundId` have type `PrimaryKey CompoundT f`? The same question applies to a bunch of other places in this module.

sashasashasasha151 replied:
I can't remember all the things that made me stop doing this. But as I remember it doesn't give you any advantages and if foreign key consists of two or more params you also can't write `PrimaryKey` for it. So, in the past I decided that it is better to not think about it as you still able to join tables

gromakovsky replied:
Okay, I didn't use this stuff before, so I don't know whether it's useful or not. Let's come to this topic later if we think that `PrimaryKey` here is better for some reason.

gromakovsky commented on backend/src/Edna/DB/Schema.hs:
```diff
@@ -1,39 +1,331 @@
+  , efMeta :: C (Nullable f) (PgJSON FileMetadata)
+  , efDescription :: C f Text
+  , efName :: C f Text
+  , efContents :: C f ByteString
```
```suggestion
  , efContents :: C f LByteString
```

Because that's what we get in `uploadFile`.

gromakovsky commented on backend/test/Test/DB/DBSpec.hs:
```diff
@@ -0,0 +1,109 @@
+  describe "Validate schema" $ modifyMaxSuccess (min 20) $ modifyMaxShrinks (const 3) $ do
+    let EdnaSchema {..} = ednaSchema
+    let insertWithConflict schema value = runInsert' $ insert schema value {-$
+          Pg.onConflict (Pg.conflictingFields primaryKey) Pg.onConflictUpdateAll-}
```
So, if it's not needed anymore, please don't forget to delete it.
